core/go/internal/sequencer/coordinator/coordinator.go

```diff
@@ -0,0 +1,620 @@
+	}
+	c.originatorNodePool = make([]string, 0)
+	c.InitializeStateMachine(State_Idle)
+	c.transactionSelector = NewTransactionSelector(ctx, c)
+	c.inFlightMutex = sync.NewCond(&sync.Mutex{})
+	c.inFlightTxns = make(map[uuid.UUID]*transaction.Transaction, c.maxDispatchAhead)
+
+	// Start event processing loop
+	go c.eventLoop(ctx)
+
+	// Start dispatch queue loop
+	go c.dispatchLoop(ctx)
+
+	// Handle loopback messages to the same node in FIFO order without blocking the event loop
+	err := transportWriter.Start(ctx)
+	if err != nil {
+		log.L(ctx).Errorf("error starting transport writer: %v", err)
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func (c *coordinator) eventLoop(ctx context.Context) {
+	for {
+		log.L(ctx).Debug("coordinator event loop waiting for next event")
+		select {
+		case event := <-c.coordinatorEvents:
+			log.L(ctx).Debugf("coordinator pulled event from the queue: %s", event.TypeString())
+			err := c.ProcessEvent(ctx, event)
+			if err != nil {
+				log.L(ctx).Errorf("error processing event: %v", err)
+			}
+		case <-c.stopEventLoop:
+		case <-c.ctx.Done():
+			log.L(ctx).Infof("coordinator event loop cancelled")
+			return
+		}
+	}
+}
+
+func (c *coordinator) dispatchLoop(ctx context.Context) {
+	dispatchedAhead := 0 // Number of transactions we've dispatched without confirming they are in the state machine's in-flight list
+
+	for {
+		log.L(ctx).Debug("coordinator dispatch loop waiting for next TX to dispatch")
+		select {
+		case tx := <-c.dispatchQueue:
+			log.L(ctx).Debugf("coordinator pulled transaction %s from the dispatch queue. In-flight count: %d, dispatched ahead: %d, max dispatch ahead: %d", tx.ID.String(), len(c.inFlightTxns), dispatchedAhead, c.maxDispatchAhead)
+
+			c.inFlightMutex.L.Lock()
+
+			// Too many in flight - wait for some to be confirmed
+			for len(c.inFlightTxns)+dispatchedAhead >= c.maxDispatchAhead {
+				c.inFlightMutex.Wait()
+			}
+
+			// Dispatch and then asynchronously update the state machine to State_Dispatched
+			log.L(ctx).Debugf("submitting transaction %s for dispatch", tx.ID.String())
+			c.readyForDispatch(ctx, tx)
+
+			// Dispatched transactions that result in a chained private transaction don't count towards max dispatch ahead
+			if tx.PreparedPrivateTransaction == nil {
+				dispatchedAhead++
+			}
+
+			// Update the TX state machine
+			c.QueueEvent(ctx, &transaction.DispatchedEvent{
+				BaseCoordinatorEvent: transaction.BaseCoordinatorEvent{
+					TransactionID: tx.ID,
+				},
+			})
+
+			// We almost never need to wait for the state machine's event loop to process the update to State_Dispatched
+			// but if we hit the max dispatch ahead limit after dispatching this transaction we do, because we can't be sure
+			// in-flight will be accurate on the next loop round
+			if len(c.inFlightTxns)+dispatchedAhead >= c.maxDispatchAhead {
+				for c.inFlightTxns[tx.ID] == nil {
+					c.inFlightMutex.Wait()
+				}
+				dispatchedAhead = 0
```

Comment:
I don't really understand what `dispatchedAhead` means here - we're arbitrarily setting it to zero here, but just because we waited for a state machine to update.

If all we're doing is propagating items from one go routine to a state machine, that's what a Go channel size is for - so we wouldn't need this layer.

So there must be something else going on here, and I've not understood it yet.